### DISCUSSION TITLE
Added a shared library but still get undefined reference

### DISCUSSION THREAD

**Discussion by memurro000** [community/community#88612] [Programming Help]
Hello there, I'm sorry if my question is silly, but I really need help

I faced a trouble trying to created a shared library

First I got a static library file `liblexbor_static.a`, then created a shared library by 
```
gcc -shared -o liblexbor.so liblexbor.a
sudo mv liblexbor.so ~/../../lib/
```

Then I created `main.c` file and compiled an executable: 
```
sudo gcc -L~/../../lib/liblexbor.so -Wall main.c -llexbor -o liblexbor.so.0.0.0
sudo mv liblexbor.so.0.0.0 ~/../../lib/
```

After that according to the manual (https://iq.opengenus.org/create-shared-library-in-cpp/) I tryed to do
```
export LD_LIBRARY_PATH=~/../../lib/liblexbor.so:$LD_LIBRARY_PATH; ./liblexbor.so.0.0.0
```
to make it available at runtime

But it does not still work:
```
g++ Main.cpp -llexbor -o Exec
./Exec
```
gives
```
g++ Main.cpp -llexbor -o Exec 
/usr/bin/ld: /tmp/ccGcCzq8.o: in function `serialize':
Main.cpp:(.text+0x70): undefined reference to `lxb_html_serialize_pretty_tree_cb'
/usr/bin/ld: /tmp/ccGcCzq8.o: in function `main':
Main.cpp:(.text+0xd6): undefined reference to `lxb_html_document_create'
/usr/bin/ld: Main.cpp:(.text+0x13c): undefined reference to `lxb_html_document_parse'
/usr/bin/ld: Main.cpp:(.text+0x1fa): undefined reference to `lxb_html_document_destroy'
collect2: error: ld returned 1 exit status
```

Please, help me fix this

Also I'd be really grateful if you helped me understand why do I `export LD_LIBRARY_PATH=~/../../lib/liblexbor.so:$LD_LIBRARY_PATH`

Maybe I should have mentioned that `main.c` consisted only of `int main() {}`

**Comment by jkoda-rsa** [comment]
It looks like you're facing a few issues with creating and using a shared library in C.

### Issues in Your Approach

1. **Library Paths and Naming**:
   - The `LD_LIBRARY_PATH` should include directories, not specific files. You've added `liblexbor.so` directly in the path, which is incorrect.
   - The naming of your shared library and executable seems to be mixed. The executable is named `liblexbor.so.0.0.0`, which is typically a convention for shared libraries, not executables.

2. **Linking Issues**:
   - The undefined reference errors suggest that the linker (`ld`) can't find the implementations of the functions from `liblexbor`.

### Steps to Fix

1. **Creating the Shared Library**:
   Your command to create a shared library from a static one seems correct:
   ```bash
      gcc -shared -o liblexbor.so liblexbor_static.a
   ```
2. **Moving the Library**:
   It's better to move your shared library to a standard library directory like /usr/local/lib:
   ```bash
     sudo mv liblexbor.so /usr/local/lib/
   ```
3. **Updating LD_LIBRARY_PATH**:
   Correct the LD_LIBRARY_PATH to include the directory:
   ```bash
      nexport LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
    ```
4. **Compiling Your Main Program**:
   Compile your main.c without specifying the path to the library:
   ```bash
       gcc -Wall main.c -llexbor -o mainExecutable
    ```
5. **Running the Program**:
   run program:
   ```bash
    ./mainExecutable
   ```
### Why Export `LD_LIBRARY_PATH`:
`LD_LIBRARY_PATH` is an environment variable for the runtime linker to find shared libraries. Adding your library path to this variable instructs the linker where to search for these libraries.

### About Your `main.c`:
If `main.c` consists solely of `int main() {}`, it won't call any functions from `liblexbor`. To properly utilize the library, you need to include the relevant headers and make calls to the library's functions as necessary.
